compose: Disable message and start new conversation buttons.

akarsh-jain-790 opened the pull request:
Implemented functionality to disable the 'Message' and 'Start new conversation' buttons when a user lacks posting permissions in the current stream context. Added tooltips for these disabled buttons, matching the behavior of the 'Send' button.

Fixes: #28410

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

https://github.com/zulip/zulip/assets/72064462/174c9cb1-f532-460e-af49-4836c70c62cc

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [X] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [X] Explains differences from previous plans (e.g., issue description).
- [X] Highlights technical choices and bugs encountered.
- [X] Calls out remaining decisions and concerns.
- [X] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [X] Each commit is a coherent idea.
- [X] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [X] Visual appearance of the changes.
- [X] Responsiveness and internationalization.
- [X] Strings and tooltips.
- [X] End-to-end functionality of buttons, interactions and flows.
- [X] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/compose_closed_ui.js:
```diff
@@ -142,22 +160,37 @@ export function update_buttons_for_private() {
+    const stream_id = filter.has_operator("stream")
+        ? stream_data.get_stream_id(filter.operands("stream")[0])
+        : undefined;
+    const stream = sub_store.get(stream_id);
```
Could some of this logic be replaced by having the caller pass in `narrow_state.stream_sub`?

akarsh-jain-790 replied:
@timabbott, I've made the requested changes and updated the PR.

alya commented:
@akarsh-jain-790 As per the PR template, please include screenshots of your changes (including tooltips) for review.

akarsh-jain-790 commented:
> @akarsh-jain-790 As per the PR template, please include screenshots of your changes (including tooltips) for review.

@alya I have updated the PR description, you can re-review now.

alya commented:
Thanks! The issue doesn't say anything about making changes in a DM context -- only changes in a stream context were requested. Did you make changes for DMs as well? I can't quite tell from the PR description.

alya commented:
We don't want to disable the "Start new conversation" button for DMs, I think, as it would make it quite difficult to start a DM with a bot or with yourself.

alya commented:
These changes are not correct. If I open the compose box in a stream where I don't have post permissions, and change the destination where the message is being sent, the UI continues to indicate that I don't have permissions. This even happens with the "New direct message" button from such a stream.

<img width="814" alt="Screenshot 2024-01-31 at 12 24 59 PM" src="https://github.com/zulip/zulip/assets/2090066/1f0818f6-c7f3-4643-915a-1adac9b5b3fd">
<img width="747" alt="Screenshot 2024-01-31 at 12 24 49 PM" src="https://github.com/zulip/zulip/assets/2090066/414f10b6-9a94-41f0-bed9-8700cf737918">

alya commented:
Additionally, there is a clear vertical alignment issue with the disabled **Start new conversation** button.
<img width="208" alt="Screenshot 2024-01-31 at 12 24 35 PM" src="https://github.com/zulip/zulip/assets/2090066/cfbf2ff9-0118-4fc3-81fc-3c11f2056c4a">

alya commented:
Please test your changes thoroughly before asking for another round of review. Requesting review for a PR with bugs you could have easily found yourself wastes valuable reviewer time and slows down the progress of the project as a whole.

akarsh-jain-790 commented:
Hi @alya ,

Apologies for the oversight in the previous changes. I've thoroughly tested the modifications now and identified and resolved the issues.

I appreciate your time. Please review the updated changes, and let me know if any further adjustments are needed.

akarsh-jain-790 commented:
Also while working on this issue, I encountered an unexpected bug. When direct messages are not permitted, the tooltip over the "compose_reply_button" is not displayed. However, upon switching to any stream, the tooltip suddenly appears, indicating that direct messages are not allowed. I've attached a screencapture and screenshots for your reference.

<img width="986" alt="Screenshot 2024-02-20 at 10 17 35 AM 1" src="https://github.com/zulip/zulip/assets/72064462/f3e31d75-8187-4d11-a283-cb6da95496ff">

https://github.com/zulip/zulip/assets/72064462/6b9f0ae9-cebb-4598-a7d7-45707ca407f2

alya commented:
Thanks! please post updated screenshots of all your changes in the PR description, including light theme and dark theme, and showing the tooltips.

alya commented:
Thanks for flagging the tooltip issue! Would you be up for posting your bug report in the #issues stream the development community?

akarsh-jain-790 commented:
> Thanks! please post updated screenshots of all your changes in the PR description, including light theme and dark theme, and showing the tooltips.

I have updated the PR description.

akarsh-jain-790 commented:
> Thanks for flagging the tooltip issue! Would you be up for posting your bug report in the #issues stream the development community?

Sure, I'll post in the #issues stream.

alya commented:
@terpimost Do we have designs for the disabled versions of compose box buttons? The colors in the screenshots above, especially for disabling "Start new conversation",  don't look great to me.

akarsh-jain-790 commented:
Hey @alya @terpimost, just checking in!  Been a while – any updates?

terpimost commented:
I didn't design any special state for disabled buttons, thinking that by default 0.5 opacity of the normal state is appropriate approach:
<img width="139" alt="image" src="https://github.com/zulip/zulip/assets/1903309/23e8f7b4-1397-46e7-a2c2-3bd2a4b8608c">

<img width="204" alt="image" src="https://github.com/zulip/zulip/assets/1903309/af3ac0a5-e6d0-495c-93db-e70947717bd3">

Please make buttons `user-select: none;` at all times. And for disabled button please make `cursor: default `

akarsh-jain-790 commented:
Hey @terpimost,

Thanks for your input. I've implemented your suggested changes and updated the PR description. Let me know if there's anything else you'd like me to adjust or if you have any further feedback!

alya commented:
Thanks! I tested the PR again, and the button does not get disabled as it should be in a search view or in All messages. Whether it's disabled should depend on the compose box state, not on what messages you're viewing.

akarsh-jain-790 commented:
> Thanks! I tested the PR again, and the button does not get disabled when it should be in a search view or in All messages. Whether it's disabled should depend on the compose box state, not on what messages you're viewing.

@alya I have updated the PR, you can review now.

<img width="1470" alt="Screenshot 2024-04-22 at 10 40 52 PM" src="https://github.com/zulip/zulip/assets/72064462/4059a676-b969-48eb-b529-904be99707a8">

alya commented:
Sure, I can take another look when tests pass.

akarsh-jain-790 commented:
> Sure, I can take another look when tests pass.

@alya Fixed the tests.

alya commented:
We should show the no permissions tooltip here, no the regular tooltip:

![Screenshot 2024-04-30 at 00 24 01](https://github.com/zulip/zulip/assets/2090066/2f3e82a5-ca49-4406-adbc-c6b9fd44d66a)

alya commented:
Otherwise looks good to me in manual testing.

akarsh-jain-790 commented:
> Otherwise looks good to me in manual testing.

I apologize for the oversight while resolving the merge conflicts. I've now fixed the issue and made all the necessary changes.

alya commented:
Thanks! @N-Shar-ma are you up for reviewing this one? I haven't retested after the latest update.

N-Shar-ma commented:
@akarsh-jain-790 On hovering over the Start new conversation button when it's disabled, I expect but don't see a `not-allowed` style cursor. Once you fix this, I'll move onto reviewing the code

alya commented:
@akarsh-jain-790 Do you plan to continue this work, or would it make sense for @N-Shar-ma to take over from here?

akarsh-jain-790 commented:
> I didn't design any special state for disabled buttons, thinking that by default 0.5 opacity of the normal state is appropriate approach: <img alt="image" width="139" src="https://private-user-images.githubusercontent.com/1903309/313412750-23e8f7b4-1397-46e7-a2c2-3bd2a4b8608c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NjQ2OTksIm5iZiI6MTcxNzU2NDM5OSwicGF0aCI6Ii8xOTAzMzA5LzMxMzQxMjc1MC0yM2U4ZjdiNC0xMzk3LTQ2ZTctYTJjMi0zYmQyYTRiODYwOGMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDVUMDUxMzE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTNmM2RhNTFiMTdiOTA4NDRiNTFhMDNmMDE1MzEwNDc1MjRjMDkzY2EzYjg5MzI1Njg0MjBlNmRlZWNhNmE5NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.qBMC7IgdwKaQuAacOO-Y4lQYb1fsyS8KnD5hkuBPBZ8">
> 
> <img alt="image" width="204" src="https://private-user-images.githubusercontent.com/1903309/313412783-af3ac0a5-e6d0-495c-93db-e70947717bd3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc1NjQ2OTksIm5iZiI6MTcxNzU2NDM5OSwicGF0aCI6Ii8xOTAzMzA5LzMxMzQxMjc4My1hZjNhYzBhNS1lNmQwLTQ5NWMtOTNkYi1lNzA5NDc3MTdiZDMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDVUMDUxMzE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjE0Njc0YjRkOTJiMTAxMWQyZDUzODY4MTY2MWQ1Mjc2ZmU2ODhmYmVkYTMxZTk5MmNiMzM5YjgyNjVhZGNlYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.6IJ3WRQlsw_2z2KnU9d1jJfAy8VmTH9-J-Wg5kFXAEA">
> Please make buttons `user-select: none;` at all times. And for disabled button please make `cursor: default `

@N-Shar-ma Thanks for pointing this out. Actually, terpimost originally suggested making disabled buttons have a default cursor style.

akarsh-jain-790 commented:
> @akarsh-jain-790 Do you plan to continue this work, or would it make sense for @N-Shar-ma to take over from here?

Sorry, I took a bit of a break from development recently after GSOC results, but now ready to jump back in.

N-Shar-ma commented:
Welcome back!

> Actually, terpimost originally suggested making disabled buttons have a default cursor style.

Oh okay. However, we still continue to have the not allowed cursor for disabled elements everywhere else, even the "Message ..." button right beside this "Start new conversation", which makes it seem odd. Just flagging this in case it's a concern (FYI @alya)

Another concern while testing this is that the tooltip on a disabled "Start new conversation" button appears instantly, and not with the usual delay

alya commented:
@terpimost Why did you suggest "And for disabled button please make `cursor: default`" here? Are you proposing that for all the close compose box buttons when they are disabled?

N-Shar-ma commented:
@terpimost just bringing your attention to the question posed in the last comment

@akarsh-jain-790 are you working on adding the expected delay to the disabled tooltip?

akarsh-jain-790 commented:
> @terpimost just bringing your attention to the question posed in the last comment
> 
> @akarsh-jain-790 are you working on adding the expected delay to the disabled tooltip?

Yes, waiting on @terpimost response on the cursor style. Then I'll push the code with the expected delay for the tooltip.

alya commented:
OK, @akarsh-jain-790 , let's leave the disabled cursor for the button on this PR. We should have a separate design discussion / issue for changing all the cursors.

akarsh-jain-790 commented:
@N-Shar-ma  I've pushed the code with the expected delay for the disabled tooltip.

N-Shar-ma commented:
@akarsh-jain-790 Please also change the cursor for the "Start new conversation" to look disabled. The delay works fine now!

alya commented:
@akarsh-jain-790 is this ready for another look from @N-Shar-ma ?

N-Shar-ma commented:
He did inform me that this is ready via DMs, it's me who needs to get around to reviewing the code now

N-Shar-ma reviewed (changes requested):
Posted some questions , since this is a bit hard to read, and the reasons for any technical decisions made aren't obvious. Adding comments might help explain some of the lines I pointed out.

Overall though it makes sense and tests well.
